File: clojure/secret-handshake/src/secret_handshake.clj
(ns secret-handshake)

(def actions ["wink" "double blink" "close your eyes" "jump"])

(defn get-actions [n]
  (keep-indexed #(when (bit-test n %1) %2) actions))

(defn commands [n] 
  (let [acts (get-actions n)]
    (if (bit-test n 4) 
      (reverse acts)
      acts)))

File: haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where
import Data.Bits (testBit)

handshake :: Int -> [String]
handshake n = 
  reverseActions $ map snd $ filter (\(idx, _) -> testBit n idx) actionsIndexed
  where 
  actionsIndexed = zip [0..] ["wink", "double blink", "close your eyes", "jump"]
  reverseActions = if testBit n 4 then reverse else id
